// nest-aws-sdk/dist/lib/aws.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var AwsSdkModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AwsSdkModule = void 0;
const common_1 = require("@nestjs/common");
const aws_service_factory_1 = require("./aws-service.factory");
const aws_service_provider_factory_1 = require("./aws-service-provider.factory");
const module_utils_1 = require("./module-utils");
const tokens_1 = require("./tokens");
let AwsSdkModule = AwsSdkModule_1 = class AwsSdkModule {
    static forRoot(options) {
        if (!options) {
            return AwsSdkModule_1.forRootAsync();
        }
        return AwsSdkModule_1.forRootAsync({
            defaultServiceOptions: options.defaultServiceOptions
                ? { useValue: options.defaultServiceOptions }
                : null,
            services: options.services,
        });
    }
    static forRootAsync(options) {
        var _a, _b;
        if (!options) {
            options = {};
        }
        if (!(options === null || options === void 0 ? void 0 : options.defaultServiceOptions)) {
            options.defaultServiceOptions = {
                useValue: {},
            };
        }
        const module = {
            global: true,
            module: AwsSdkModule_1,
            imports: [],
            providers: [aws_service_factory_1.AwsServiceFactory],
            exports: [aws_service_factory_1.AwsServiceFactory],
        };
        const serviceOptionsProvider = (0, module_utils_1.createExportableProvider)(tokens_1.AWS_SERVICE_CONFIGURATION_OPTIONS_FACTORY_TOKEN, options === null || options === void 0 ? void 0 : options.defaultServiceOptions);
        if (serviceOptionsProvider.imports.length) {
            serviceOptionsProvider.imports.forEach(i => module.imports.push(i));
        }
        if (serviceOptionsProvider.exports.length) {
            serviceOptionsProvider.exports.forEach(i => module.exports.push(i));
        }
        if (serviceOptionsProvider.provider) {
            module.providers.push(serviceOptionsProvider.provider);
        }
        const serviceProviders = (_b = (_a = options.services) === null || _a === void 0 ? void 0 : _a.map(s => (0, aws_service_provider_factory_1.createAwsServiceProvider)(s))) !== null && _b !== void 0 ? _b : [];
        if (serviceProviders.length) {
            serviceProviders.forEach(sp => {
                module.providers.push(sp);
                module.exports.push(sp.provide);
            });
        }
        return module;
    }
    static forFeatures(services) {
        var _a;
        const serviceProviders = (_a = services === null || services === void 0 ? void 0 : services.map(s => (0, aws_service_provider_factory_1.createAwsServiceProvider)(s))) !== null && _a !== void 0 ? _a : [];
        const module = {
            module: AwsSdkModule_1,
            providers: [...serviceProviders, aws_service_factory_1.AwsServiceFactory],
            exports: [...serviceProviders, aws_service_factory_1.AwsServiceFactory],
        };
        return module;
    }
};
AwsSdkModule = AwsSdkModule_1 = __decorate([
    (0, common_1.Module)({})
], AwsSdkModule);
exports.AwsSdkModule = AwsSdkModule;
